src/fillalgebra.jl

```diff
@@ -399,7 +399,7 @@ dot(a::AbstractVector, b::AbstractFillVector) = _fill_dot_rev(a, b)
 function dot(u::AbstractVector, E::Eye, v::AbstractVector)
     length(u) == size(E,1) && length(v) == size(E,2) ||
         throw(DimensionMismatch("dot product arguments have dimensions $(length(u))×$(size(E))×$(length(v))"))
-    dot(u, v)
+    one(eltype(E)) * dot(u, v)
```

Comment:
Doesn't this do an unnecessary multiplication? I would think a `convert` would be better

Reply:
The problem with the `convert` approach is that we need to know the final type, but `promote_type` may produce a wider result than the actual type. E.g. if `promote_type` produces `Any`, the `convert(Any, x)` would not change the type of `x`. `LinearAlgebra` gets around this by using `Base.promote_op`, but that's an internal function that's not recommended.
I guess we may just use `promote_type` here, which should work for `Number`s. I had used the multiplication here to avoid the issue altogether, and ensure that the correct type is returned.

Reply:
This would avoid the unnecessary multiplication:
```suggestion
d = dot(u,v)
T = eltype(one(eltype(E)) * d)
convert(T, d)
```

Reply:
Isn't the multiplication still there in `eltype(one(eltype(E)) * d)`?

Reply:
No it gets compiled out when multiplication is pure:
```julia
julia> function foo(E, d)
       eltype(one(eltype(E)) * d)
       end
foo (generic function with 1 method)

julia> foo(2.0, 3)
Float64

julia> @code_warntype foo(2.0, 3)
MethodInstance for foo(::Float64, ::Int64)
  from foo(E, d) @ Main REPL[1]:1
Arguments
  #self#::Core.Const(foo)
  E::Float64
  d::Int64
Body::Type{Float64}
1 ─ %1 = Main.eltype(E)::Core.Const(Float64)
│   %2 = Main.one(%1)::Core.Const(1.0)
│   %3 = (%2 * d)::Float64
│   %4 = Main.eltype(%3)::Core.Const(Float64)
└──      return %4

```

Reply:
Actually that doesn't show it ...  but I'm pretty sure it does get compiled out

Reply:
Do you mean `typeof` instead of `eltype` on line 403?